core/src/main/java/io/methvin/watcher/DirectoryWatcher.java

```diff
@@ -35,14 +38,15 @@
 import static java.nio.file.LinkOption.NOFOLLOW_LINKS;
 import static java.nio.file.StandardWatchEventKinds.*;
 
-public class DirectoryWatcher {
+public class DirectoryWatcher<C> {
 
   /**
    * A builder for a {@link DirectoryWatcher}. Use {@code DirectoryWatcher.builder()} to get a new
    * instance.
    */
-  public static final class Builder {
+  public static final class Builder<C> {
     private List<Path> paths = Collections.emptyList();
+    private Map<Path, C> contexts = new HashMap<>();
```

Comment:
ok, I think I can change it to do that. It amounts to much the same code, there will still be a context - but not user provided. Just instead of C, it'll be the root Path. This gets passed as an argument, and then up to the user to map back. I'll update the PR for this over the weekend.

Reply:
This is now fixed, the context is now internalised and just paths to the original provided paths.